### PR TITLE
chore: adding NetConfig test suite

### LINKED ISSUE #1540: chore: improve test coverage on NetConfig generation
### Background
In #1512, the waku node network config was abstracted out.
Test coverage on address generation is low at the moment.

### Details
Tests for the network config generation to improve coverage to detect regressions.

### Acceptance criteria
- [ ] Identify common network configs
- [ ] Tests for the same

cc: @jm-clius

### PR DESCRIPTION
**Pull request by gabrielmer**
# Description
Added test suite to cover common NetConfig settings

# Changes

- [x] Created new file `./tests/test_waku_netconfig.nim` with a NetConfig test suite
- [x] Moved `defaultTestWakuNodeConf` from `test_app.nim` to `external_config.nim`
- [x] Updated `defaultTestWakuNodeConf`to use our current default `listenAddress` `0.0.0.0` and added more default parameters

## Issue

closes #1540

### CODE REVIEW

File: tests/testlib/wakunode.nim
Comment (gabrielmer) [inline tests/testlib/wakunode.nim:17]: When importing `defaultTestWakuNodeConf` from `test_app` into other test suites, the `test_app` suite was being re-run as part of the suites importing it. For easier `defaultTestWakuNodeConf`'s reusability, moved the function to `external_config`, mimicking  `newTestWakuNode`, which is reused by many test suites and is defined in `wakunode.nim`

File: tests/test_waku_netconfig.nim
Comment (gabrielmer) [inline tests/test_waku_netconfig.nim:205]: Note that in order to configure WS using an `extIp`, the published port isn't `extPort` but `conf.websocketPort`. For websockets we bind to and announce the same port number.

File: tests/test_waku_netconfig.nim
Comment (jm-clius) [inline tests/test_waku_netconfig.nim:66]: Ditto in terms of indentation and for other tests (I think they should be single indented to be considered part of "Waku Netconfig" `suite`

**Review by jm-clius — COMMENTED** [review]
Looks good and covers quite a lot! Thanks for this. Much better than what we have (which was noting 😅 ). Feel free to mark as ready for review once you think it can be merged.

File: tests/test_waku_netconfig.nim
Comment (gabrielmer) [reply tests/test_waku_netconfig.nim]: Good catch! For some reason the tests were still being run as part of the test suite with that indentation, which is pretty interesting. Just pushed a fix for it - thank you!

File: apps/wakunode2/external_config.nim
Comment (jm-clius) [inline apps/wakunode2/external_config.nim:568]: Why has this been moved from `test_app.nim`? I wouldn't move any test code into "production" code.

File: tests/testlib/wakunode.nim
Comment (jm-clius) [reply tests/testlib/wakunode.nim:17]: Ah, made a comment about this elsewhere before seeing this. `wakunode.nim` is a `testlib` - in other words, it's designed to be a utility module for functions useful in tests. `newTestWakuNode` `defaultTestWakuNodeConf` etc. shouldn't be defined in any "production" modules (test code should remain completely separate. The way to prevent `test_app` from being run multiple times for each import is to move `defaultTestWakuNodeConf` to a test helper module. I'd suggest just adding it to `testlib/wakunode.nim` and importing in both `test_app` and `test_waku_netconfig`

**Review by jm-clius — APPROVED** [review]
LGTM. `defaultTestWakuNodeConf` should not be defined in a "production" module llike `external_config.nim`, but once this has been moved I'll be happy for the suite to be merged. No need to re-request review from my side. Great job!

**Comment by gabrielmer** [comment]
> LGTM. `defaultTestWakuNodeConf` should not be defined in a "production" module llike `external_config.nim`, but once this has been moved I'll be happy for the suite to be merged. No need to re-request review from my side. Great job!

Ooh now everything makes much more sense! I didn't notice that `wakunode.nim` was a `testlib` 😶
Will move `defaultTestWakuNodeConf`  to `wakunode.nim` then :)

Thanks so much!

File: tests/test_waku_netconfig.nim
Comment (Ivansete-status) [inline tests/test_waku_netconfig.nim:58]: I'd always apply the next technique for such cases:

```suggestion
    assert netConfigRes.isOk(), $netConfigRes.error
```

That applies elsewhere.

File: tests/test_waku_netconfig.nim
Comment (Ivansete-status) [inline tests/test_waku_netconfig.nim:13]: Sorry, this is a personal obsession hehe. If possible I recommend always to keep one empty line max.
```suggestion
```

File: waku/node/config.nim
Comment (Ivansete-status) [inline waku/node/config.nim:39]: Isn't there a way to make the tests without changing the scope of that?
If possible, I recommend only testing the public procs and don't change the scope of private ones. If you still need to use these private procs you may need to `include` the `../../waku/node/config,` in the test, instead of `import`.

**Review by Ivansete-status — CHANGES_REQUESTED** [review]
Thanks for that! You created many more test cases than I could imagine :1st_place_medal: 
Nevertheless, there are a few things that need to be revisited :)
It is especially important the comment about `apps/wakunode2/external_config.nim`. I'd prevent changing it.
Cheers!

**Review by SionoiS — APPROVED** [review]
No further comments than what was already said.

I just feel like we could use the type system in cases like this to enforce some "rules".

My 0.02$

File: tests/test_waku_netconfig.nim
Comment (gabrielmer) [reply tests/test_waku_netconfig.nim]: Agree that it's a better technique! Done :)

File: waku/node/config.nim
Comment (gabrielmer) [reply waku/node/config.nim]: Makes sense! I changed it to `import` instead. I noticed that for some reason really few test files use `import` - I wonder how many private procs have been turned into public for testing, probably a lot. Anyway, I agree on the principle of making the least amount possible of changes in prod code that aren't related to prod logic.

File: tests/test_waku_netconfig.nim
Comment (gabrielmer) [reply tests/test_waku_netconfig.nim:12]: Hahaha done, deleted that line and also a couple more instances in the PR with extra empty lines :)

**Comment by gabrielmer** [comment]
> No further comments than what was already said.
> 
> I just feel like we could use the type system in cases like this to enforce some "rules".
> 
> My 0.02$

I'm not sure I fully understand the idea. Is it about setting explicitly the type when declaring variables instead of having the types being automatically inferred so we enforce the return types of procs?

Or what kind of rules are you thinking of?

**Review by Ivansete-status — APPROVED** [review]
LGTM! Thanks for it! :100:

**Comment by SionoiS** [comment]
> > No further comments than what was already said.
> > I just feel like we could use the type system in cases like this to enforce some "rules".
> > My 0.02$
> 
> I'm not sure I fully understand the idea. Is it about setting explicitly the type when declaring variables instead of having the types being automatically inferred so we enforce the return types of procs?
> 
> Or what kind of rules are you thinking of?

Might not be practical to do this here but the type system can help enforcing requirement like
`AnnouncedAddresses contains only bind address when no external addresses are provided`

For example you could use some type to represent the state.
- `only bind address when no external addresses are provided`
- `only external address if extIp/Port are provided`

**Pull request merged**
